### ISSUE TITLE
Rule suggestion: prefer-parameter-properties

### ISSUE THREAD

**Issue by danvk** [palantir/tslint#2638]
There's a [`no-parameter-properties`][2] rule, but what if you like them? How about a rule that flags code like this:

```ts
class Foo {
  foo: number;
  constructor(foo: number) {
    this.foo = foo;
  }
}
```

in favor of this:

```ts
class Foo {
  constructor(public foo: number) {
  }
}
```

Another option would be to replace `no-parameter-properties` with a more general `parameter-properties` rule that lets you say how you feel about them.

See [parameter properties][1] in the TS handbook.

[1]: https://www.typescriptlang.org/docs/handbook/classes.html#parameter-properties
[2]: https://palantir.github.io/tslint/rules/no-parameter-properties/

**Comment by adidahiya** [comment]
Sure, that sounds good. I would like it to be one rule though, so let's add `parameter-properties` with options `"always" | "never"` and deprecate `no-parameter-properties`.

**Comment by codingthat** [comment]
What was the motivation for this rule in the first place?

**Comment by adidahiya** [comment]
@codingthat mostly just consistency with other languages like Java for full-stack developers.

**Comment by codingthat** [comment]
@adidahiya That being the case, I'm even happier to see the rule go neutral.  :)

**Comment by adidahiya** [comment]
I mean, it _is_ harder to see the list of members at a glance when you use parameter properties. New language features almost always add more cognitive load.

**Comment by codingthat** [comment]
Depends on your formatting, maybe :)  That's a good point to consider though.

**Comment by JoshuaKGoldberg** [comment]
Note: per #4534, this issue will be closed in less than a month if no PR is sent to add the rule. If you _really_ need the rule, custom rules are always an option and can be maintained outside this repo!

**Comment by JoshuaKGoldberg** [comment]
### 💀 _**It's time!**_ 💀

**TSLint is being deprecated** and no longer accepting pull requests for new rules. See #4534. 😱

If you'd like to see this rule implemented, you have two choices:
* **Recommended:** Use [typescript-eslint](https://typescript-eslint.io) to lint your TypeScript code with ESLint ✅
* **Legacy:** Write it as a [custom TSLint rule](https://palantir.github.io/tslint/develop/custom-rules/) 🤷‍♂️

👋 It was a pleasure open sourcing with you!

_If you believe this message was posted here in error, please comment so we can re-open the issue!_
